tests/test_minilinq.py

```diff
@@ -88,6 +86,31 @@ def test_eval_auto_id_reference_nested(self):
         #   Reference('$.foo.id'):
         #       '1.bid' -> 'bid'
 
+    def test_flatmap_value_or_root(self):
+        """Low level test case for 'value-or-root' use case"""
+        env = BuiltInEnv() | JsonPathEnv({})
+
+        data = [
+            {"id": 1, "foo": {'id': 'bid', 'name': 'zip'}, "bar": [{'baz': 'a1'}, {'baz': 'a2', 'id': 'bazzer'}]},
+            {"id": 2, "foo": {'id': 'bid', 'name': 'zap'}, "bar": []},
+            {"id": 3, "foo": {'id': 'bid', 'name': 'mip'}, "bar": {}},
+            # {"id": 4, "foo": {'id': 'bid', 'name': 'map'}, "bar": None},  # fails with TypeError from jsonpath
+            {"id": 5, "foo": {'id': 'bid', 'name': 'mop'}},
+        ]
+        value_or_root = get_value_or_root_expression('bar.[*]')
+        flatmap = FlatMap(source=Literal(data), body=value_or_root)
+        mmap = Map(source=flatmap, body=List([
+            Reference("id"), Reference('baz'), Reference('$.id'), Reference('$.foo.id'), Reference('$.foo.name')
```

Comment:
would it be worth adding a test case where `baz` exists on the parent (and the child is empty)? Is the behavior there that it would use it or that it would remain empty if it has fallen back to the parent emit (my guess is the desired behavior is to be empty)

Reply:
good question, it would actually take the value from the parent:
```python
data = {"id": 6, "foo": {'id': 'bid', 'name': 'mop'}, "baz": "root_bazz"}
# id, baz, $.id, $.foo.id, $.foo.name
result = ['6', 'root_bazz', '6', '6.bid', 'mop']
```

That definitely isn't great. I have no idea how to work around that. I might be able to hack it by setting a special var in the env if we're in the case where we're using the root doc and then change the JsonPathEnv to only allow 'root' expressions.

Reply:
Hmm - I guess it's not possible to emit something like `{'$': root}` instead of the root document itself? (not sure if that even makes sense)

I don't know that it's an important enough edge-case to worry about, though does seem worth at least documenting the "expected" behavior in a test

Reply:
Not loving this but.... 8ac159f